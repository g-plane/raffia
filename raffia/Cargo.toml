[package]
name = "raffia"
version = "0.1.0"
edition = "2021"
authors = ["Pig Fang <g-plane@hotmail.com>"]
description = "One parser for CSS/SCSS/Sass/Less."
repository = "https://github.com/g-plane/raffia"
license = "MIT"

[dependencies]
raffia_macro = { path = "../raffia_macro", version = "0.1" }
serde = { version = "1.0", features = ["derive"], optional = true }
smallvec = { version = "1.9", features = ["const_generics", "serde"] }

[dev-dependencies]
codespan-reporting = "0.11"
criterion = "0.3"
insta = { version = "1.19", features = ["glob", "ron"] }
lightningcss = "1.0.0-alpha.35"
serde_json = "1.0"
swc_common = "0.29.0"
swc_css_ast = "0.114.0"
swc_css_parser = "0.123.0"

[features]
serialize = ["serde"]

[[bench]]
name = "parse_selector"
harness = false

[[bench]]
name = "parse_stylesheet"
harness = false
