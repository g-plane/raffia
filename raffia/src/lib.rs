//! Raffia is a parser can parse CSS, SCSS, Sass (indented syntax) and Less.
//!
//! ## Basic Usage
//!
//! This crate provides a simple API to get started.
//!
//! First, create a parser, give it the source code and specify the syntax,
//! then call the [`parse`](Parser::parse) method:
//!
//! ```rust
//! use raffia::{ast::Stylesheet, Parser, Syntax};
//!
//! let mut parser = Parser::new("a {}", Syntax::Css); // syntax can also be `Scss`, `Sass` or `Less`
//! let result = parser.parse::<Stylesheet>();
//! match result {
//!     Ok(ast) => {
//!         // parsed successfully
//!         println!("{:#?}", ast);
//!     }
//!     Err(error) => {
//!         // it failed, error message and position can be accessed via `error`
//!         println!("{:#?}", error);
//!     }
//! }
//! ```
//!
//! ## Advanced Usage
//!
//! ### More Parser Options
//!
//! If you need to control more parser options, you can use [`ParserBuilder`].
//!
//! For example, to collect comments:
//!
//! ```rust
//! use raffia::ParserBuilder;
//!
//! let mut comments = vec![];
//! let builder = ParserBuilder::new("/* comment */ a {}").comments(&mut comments);
//! let mut parser = builder.build();
//! ```
//!
//! By default, syntax is CSS when using parser builder. You can customize it:
//!
//! ```rust
//! use raffia::{ParserBuilder, Syntax};
//!
//! let builder = ParserBuilder::new("a {}").syntax(Syntax::Scss);
//! ```
//!
//! ### Parse Partial Structure
//!
//! Sometimes you don't want to parse a full stylesheet.
//! Say you only need to parse a qualified rule or even a single declaration.
//! All you need to do is to update the generics of the [`parse`](Parser::parse) method.
//!
//! ```rust
//! use raffia::{ast::QualifiedRule, Parser, Syntax};
//!
//! let mut parser = Parser::new("a {}", Syntax::Css);
//! parser.parse::<QualifiedRule>();
//! ```
//!
//! and
//!
//! ```rust
//! use raffia::{ast::Declaration, Parser, Syntax};
//!
//! let mut parser = Parser::new("color: green", Syntax::Css);
//! parser.parse::<Declaration>();
//! ```
//!
//! Not all AST nodes support the usage above;
//! technically, those nodes that implement [`Parse`] trait are supported.
//!
//! ### Retrieve Recoverable Errors
//!
//! There may be some recoverable errors which doesn't affect on producing AST.
//! To retrieve those errors, use [`recoverable_errors`](Parser::recoverable_errors).
//!
//! ```rust
//! use raffia::{ast::Stylesheet, Parser, Syntax};
//!
//! let mut parser = Parser::new("@keyframes kf { invalid {} }", Syntax::Css);
//! let result = parser.parse::<Stylesheet>();
//! assert!(result.is_ok());
//! println!("{:?}", parser.recoverable_errors());
//! ```
//!
//! ## Serialization
//!
//! Produced AST can be serialized by Serde, but this feature is disabled by default.
//! You need to enable feature `serialize` manually:
//!
//! ```toml
//! raffia = { version = "*", features = ["serialize"] }
//! ```
//!
//! Then you can pass AST to Serde.
//!
//! Note that Raffia only supports serialization. Deserialization isn't supported.

pub use config::{ParserOptions, Syntax};
pub use parser::{Parse, Parser, ParserBuilder};
pub use pos::{Span, Spanned};
pub use span_ignored_eq::SpanIgnoredEq;
pub use tokenizer::token;

pub mod ast;
mod config;
pub mod error;
mod parser;
pub mod pos;
mod span_ignored_eq;
mod tokenizer;
mod util;
