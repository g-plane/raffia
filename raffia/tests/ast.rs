use codespan_reporting::{
    diagnostic::{Diagnostic, Label},
    files::SimpleFile,
    term::{self, termcolor::Buffer},
};
use insta::{assert_ron_snapshot, glob, Settings};
use raffia::{ast::Stylesheet, Parser, Syntax};
use std::fs;

#[test]
fn ast_snapshot() {
    glob!("ast/**/*.{css,scss,sass,less}", |path| {
        let code = fs::read_to_string(path).unwrap();
        let syntax = match path.extension().unwrap().to_str().unwrap() {
            "css" => Syntax::Css,
            "scss" => Syntax::Scss,
            "sass" => Syntax::Sass,
            "less" => Syntax::Less,
            _ => unreachable!("unknown file extension"),
        };
        let mut parser = Parser::new(&code, syntax);
        let ast = match parser.parse::<Stylesheet>() {
            Ok(ast) => {
                let recoverable_errors = parser.recoverable_errors();
                assert!(
                    recoverable_errors.is_empty(),
                    "'{}' has recoverable errors: {recoverable_errors:?}",
                    path.file_name().unwrap().to_str().unwrap(),
                );
                ast
            }
            Err(error) => {
                let file = SimpleFile::new(path.file_name().unwrap().to_str().unwrap(), &code);
                let diagnostic = Diagnostic::error()
                    .with_message(error.kind.to_string())
                    .with_labels(vec![Label::primary((), error.span.start..error.span.end)]);
                let mut buffer = Buffer::ansi();
                let config = term::Config::default();
                term::emit(&mut buffer, &config, &file, &diagnostic).unwrap();
                panic!("\n{}", String::from_utf8(buffer.into_inner()).unwrap());
            }
        };

        let mut settings = Settings::clone_current();
        settings.set_snapshot_path(path.parent().unwrap());
        settings.remove_snapshot_suffix();
        settings.set_prepend_module_to_snapshot(false);
        settings.remove_input_file();
        settings.set_omit_expression(true);
        settings.remove_input_file();
        settings.remove_info();
        settings.bind(|| {
            let name = path.file_stem().unwrap().to_str().unwrap();
            assert_ron_snapshot!(name, ast);
        });
    });
}
