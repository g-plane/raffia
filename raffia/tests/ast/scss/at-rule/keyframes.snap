---
source: raffia/tests/ast.rs
---
Stylesheet(
  type: "Stylesheet",
  statements: [
    SassMixinAtRule(
      type: "SassMixinAtRule",
      name: Ident(
        type: "Ident",
        name: "my-keyframes",
        raw: "my-keyframes",
        span: Span(
          start: 7,
          end: 19,
        ),
      ),
      parameters: Some([]),
      arbitraryParameter: None,
      body: SimpleBlock(
        type: "SimpleBlock",
        statements: [
          AtRule(
            type: "AtRule",
            name: Ident(
              type: "Ident",
              name: "-webkit-keyframes",
              raw: "-webkit-keyframes",
              span: Span(
                start: 27,
                end: 44,
              ),
            ),
            prelude: Some(Ident(
              type: "Ident",
              name: "kf",
              raw: "kf",
              span: Span(
                start: 45,
                end: 47,
              ),
            )),
            block: Some(SimpleBlock(
              type: "SimpleBlock",
              statements: [
                SassContentAtRule(
                  type: "SassContentAtRule",
                  arguments: None,
                  span: Span(
                    start: 54,
                    end: 62,
                  ),
                ),
              ],
              span: Span(
                start: 48,
                end: 67,
              ),
            )),
            span: Span(
              start: 26,
              end: 67,
            ),
          ),
          AtRule(
            type: "AtRule",
            name: Ident(
              type: "Ident",
              name: "-ms-keyframes",
              raw: "-ms-keyframes",
              span: Span(
                start: 71,
                end: 84,
              ),
            ),
            prelude: Some(Ident(
              type: "Ident",
              name: "kf",
              raw: "kf",
              span: Span(
                start: 85,
                end: 87,
              ),
            )),
            block: Some(SimpleBlock(
              type: "SimpleBlock",
              statements: [
                SassContentAtRule(
                  type: "SassContentAtRule",
                  arguments: None,
                  span: Span(
                    start: 94,
                    end: 102,
                  ),
                ),
              ],
              span: Span(
                start: 88,
                end: 106,
              ),
            )),
            span: Span(
              start: 70,
              end: 106,
            ),
          ),
          AtRule(
            type: "AtRule",
            name: Ident(
              type: "Ident",
              name: "-moz-keyframes",
              raw: "-moz-keyframes",
              span: Span(
                start: 110,
                end: 124,
              ),
            ),
            prelude: Some(Ident(
              type: "Ident",
              name: "kf",
              raw: "kf",
              span: Span(
                start: 125,
                end: 127,
              ),
            )),
            block: Some(SimpleBlock(
              type: "SimpleBlock",
              statements: [
                SassContentAtRule(
                  type: "SassContentAtRule",
                  arguments: None,
                  span: Span(
                    start: 134,
                    end: 142,
                  ),
                ),
              ],
              span: Span(
                start: 128,
                end: 147,
              ),
            )),
            span: Span(
              start: 109,
              end: 147,
            ),
          ),
          AtRule(
            type: "AtRule",
            name: Ident(
              type: "Ident",
              name: "-o-keyframes",
              raw: "-o-keyframes",
              span: Span(
                start: 151,
                end: 163,
              ),
            ),
            prelude: Some(Ident(
              type: "Ident",
              name: "kf",
              raw: "kf",
              span: Span(
                start: 164,
                end: 166,
              ),
            )),
            block: Some(SimpleBlock(
              type: "SimpleBlock",
              statements: [
                SassContentAtRule(
                  type: "SassContentAtRule",
                  arguments: None,
                  span: Span(
                    start: 173,
                    end: 181,
                  ),
                ),
              ],
              span: Span(
                start: 167,
                end: 185,
              ),
            )),
            span: Span(
              start: 150,
              end: 185,
            ),
          ),
          AtRule(
            type: "AtRule",
            name: Ident(
              type: "Ident",
              name: "keyframes",
              raw: "keyframes",
              span: Span(
                start: 189,
                end: 198,
              ),
            ),
            prelude: Some(Ident(
              type: "Ident",
              name: "kf",
              raw: "kf",
              span: Span(
                start: 199,
                end: 201,
              ),
            )),
            block: Some(SimpleBlock(
              type: "SimpleBlock",
              statements: [
                SassContentAtRule(
                  type: "SassContentAtRule",
                  arguments: None,
                  span: Span(
                    start: 208,
                    end: 216,
                  ),
                ),
              ],
              span: Span(
                start: 202,
                end: 221,
              ),
            )),
            span: Span(
              start: 188,
              end: 221,
            ),
          ),
          AtRule(
            type: "AtRule",
            name: Ident(
              type: "Ident",
              name: "keyframes",
              raw: "keyframes",
              span: Span(
                start: 225,
                end: 234,
              ),
            ),
            prelude: Some(Ident(
              type: "Ident",
              name: "kf",
              raw: "kf",
              span: Span(
                start: 235,
                end: 237,
              ),
            )),
            block: Some(SimpleBlock(
              type: "SimpleBlock",
              statements: [
                SassVariableDeclaration(
                  type: "SassVariableDeclaration",
                  namespace: None,
                  name: SassVariable(
                    type: "SassVariable",
                    name: Ident(
                      type: "Ident",
                      name: "var",
                      raw: "var",
                      span: Span(
                        start: 245,
                        end: 248,
                      ),
                    ),
                    span: Span(
                      start: 244,
                      end: 248,
                    ),
                  ),
                  value: Ident(
                    type: "Ident",
                    name: "true",
                    raw: "true",
                    span: Span(
                      start: 250,
                      end: 254,
                    ),
                  ),
                  overridable: false,
                  forceGlobal: false,
                  span: Span(
                    start: 244,
                    end: 255,
                  ),
                ),
                SassIfAtRule(
                  type: "SassIfAtRule",
                  ifClause: SassConditionalClause(
                    type: "SassConditionalClause",
                    condition: SassVariable(
                      type: "SassVariable",
                      name: Ident(
                        type: "Ident",
                        name: "var",
                        raw: "var",
                        span: Span(
                          start: 265,
                          end: 268,
                        ),
                      ),
                      span: Span(
                        start: 264,
                        end: 268,
                      ),
                    ),
                    block: SimpleBlock(
                      type: "SimpleBlock",
                      statements: [
                        KeyframeBlock(
                          type: "KeyframeBlock",
                          selectors: [
                            Percentage(
                              type: "Percentage",
                              value: Number(
                                type: "Number",
                                value: 0.0,
                                raw: "0",
                                span: Span(
                                  start: 277,
                                  end: 278,
                                ),
                              ),
                              span: Span(
                                start: 277,
                                end: 279,
                              ),
                            ),
                          ],
                          block: SimpleBlock(
                            type: "SimpleBlock",
                            statements: [
                              Declaration(
                                type: "Declaration",
                                name: Ident(
                                  type: "Ident",
                                  name: "color",
                                  raw: "color",
                                  span: Span(
                                    start: 290,
                                    end: 295,
                                  ),
                                ),
                                value: [
                                  Ident(
                                    type: "Ident",
                                    name: "red",
                                    raw: "red",
                                    span: Span(
                                      start: 297,
                                      end: 300,
                                    ),
                                  ),
                                ],
                                important: None,
                                lessPropertyMerge: None,
                                span: Span(
                                  start: 290,
                                  end: 300,
                                ),
                              ),
                            ],
                            span: Span(
                              start: 280,
                              end: 309,
                            ),
                          ),
                          span: Span(
                            start: 277,
                            end: 309,
                          ),
                        ),
                        KeyframeBlock(
                          type: "KeyframeBlock",
                          selectors: [
                            Percentage(
                              type: "Percentage",
                              value: Number(
                                type: "Number",
                                value: 100.0,
                                raw: "100",
                                span: Span(
                                  start: 316,
                                  end: 319,
                                ),
                              ),
                              span: Span(
                                start: 316,
                                end: 320,
                              ),
                            ),
                          ],
                          block: SimpleBlock(
                            type: "SimpleBlock",
                            statements: [
                              Declaration(
                                type: "Declaration",
                                name: Ident(
                                  type: "Ident",
                                  name: "color",
                                  raw: "color",
                                  span: Span(
                                    start: 331,
                                    end: 336,
                                  ),
                                ),
                                value: [
                                  Ident(
                                    type: "Ident",
                                    name: "blue",
                                    raw: "blue",
                                    span: Span(
                                      start: 338,
                                      end: 342,
                                    ),
                                  ),
                                ],
                                important: None,
                                lessPropertyMerge: None,
                                span: Span(
                                  start: 331,
                                  end: 342,
                                ),
                              ),
                            ],
                            span: Span(
                              start: 321,
                              end: 351,
                            ),
                          ),
                          span: Span(
                            start: 316,
                            end: 351,
                          ),
                        ),
                      ],
                      span: Span(
                        start: 269,
                        end: 357,
                      ),
                    ),
                    span: Span(
                      start: 264,
                      end: 357,
                    ),
                  ),
                  elseIfClauses: [],
                  elseClause: Some(SimpleBlock(
                    type: "SimpleBlock",
                    statements: [
                      KeyframeBlock(
                        type: "KeyframeBlock",
                        selectors: [
                          Ident(
                            type: "Ident",
                            name: "from",
                            raw: "from",
                            span: Span(
                              start: 372,
                              end: 376,
                            ),
                          ),
                        ],
                        block: SimpleBlock(
                          type: "SimpleBlock",
                          statements: [
                            Declaration(
                              type: "Declaration",
                              name: Ident(
                                type: "Ident",
                                name: "color",
                                raw: "color",
                                span: Span(
                                  start: 387,
                                  end: 392,
                                ),
                              ),
                              value: [
                                Ident(
                                  type: "Ident",
                                  name: "blue",
                                  raw: "blue",
                                  span: Span(
                                    start: 394,
                                    end: 398,
                                  ),
                                ),
                              ],
                              important: None,
                              lessPropertyMerge: None,
                              span: Span(
                                start: 387,
                                end: 398,
                              ),
                            ),
                          ],
                          span: Span(
                            start: 377,
                            end: 407,
                          ),
                        ),
                        span: Span(
                          start: 372,
                          end: 407,
                        ),
                      ),
                      KeyframeBlock(
                        type: "KeyframeBlock",
                        selectors: [
                          Ident(
                            type: "Ident",
                            name: "to",
                            raw: "to",
                            span: Span(
                              start: 414,
                              end: 416,
                            ),
                          ),
                        ],
                        block: SimpleBlock(
                          type: "SimpleBlock",
                          statements: [
                            Declaration(
                              type: "Declaration",
                              name: Ident(
                                type: "Ident",
                                name: "color",
                                raw: "color",
                                span: Span(
                                  start: 427,
                                  end: 432,
                                ),
                              ),
                              value: [
                                Ident(
                                  type: "Ident",
                                  name: "red",
                                  raw: "red",
                                  span: Span(
                                    start: 434,
                                    end: 437,
                                  ),
                                ),
                              ],
                              important: None,
                              lessPropertyMerge: None,
                              span: Span(
                                start: 427,
                                end: 437,
                              ),
                            ),
                          ],
                          span: Span(
                            start: 417,
                            end: 446,
                          ),
                        ),
                        span: Span(
                          start: 414,
                          end: 446,
                        ),
                      ),
                    ],
                    span: Span(
                      start: 364,
                      end: 452,
                    ),
                  )),
                  span: Span(
                    start: 260,
                    end: 452,
                  ),
                ),
              ],
              span: Span(
                start: 238,
                end: 456,
              ),
            )),
            span: Span(
              start: 224,
              end: 456,
            ),
          ),
        ],
        span: Span(
          start: 22,
          end: 458,
        ),
      ),
      span: Span(
        start: 0,
        end: 458,
      ),
    ),
  ],
  span: Span(
    start: 0,
    end: 459,
  ),
)
